### PR TITLE
CUMULUS-1799: Add more new packages

### PR DESCRIPTION
**Pull request by markdboyd**
- Adds @cumulus/launchpad-auth package and refactored code to use it
- Added @cumulus/collection/config-store package and refactored code to use it

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -23,14 +26,15 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - **CUMULUS-1799** - Deprecated the following code. For cases where the code was moved into another package, the new code location is noted:
   - `@cumulus/aws-client/StepFunctions.fromSfnExecutionName()`
   - `@cumulus/aws-client/StepFunctions.toSfnExecutionName()`
-  - `@cumulus/aws-client/StepFunctions.getExecutionArn()` -> `@cumulus/message/executions.buildExecutionArn()`
-  - `@cumulus/aws-client/StepFunctions.getExecutionUrl()` -> `@cumulus/message/executions.getExecutionUrlFromArn()`
-  - `@cumulus/aws-client/StepFunctions.getStateMachineArn()` -> `@cumulus/message/executions.getStateMachineArnFromExecutionArn()`
+  - `@cumulus/aws-client/StepFunctions.getExecutionArn()` -> `@cumulus/message/Executions.buildExecutionArn()`
```
Comment (Jkovarik) [inline CHANGELOG.md:29]: I'm assuming this is just an unrelated doc change?

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,10 +8,13 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Added
 
+- **CUMULUS-1787**
+  - Added `collections/active` endpoint for returning collections with active granules in `@cumulus/api`
```
Comment (Jkovarik) [inline CHANGELOG.md:12]: Is this intentional given the PR is for 1799?

File: packages/collection-config-store/tests/test-collection-config-store.js [hunk 3]
```diff
@@ -0,0 +1,207 @@
+'use strict';
```
Comment (Jkovarik) [inline packages/collection-config-store/tests/test-collection-config-store.js:1]: Nit/git thought/rando stream of consciousness: 

I wonder if we should  have git moved these, then added a new copy to the soon-to-be-deprecated tests to preserve line/commit history and/or for ease of review?        

This file *does* differ from ../cumulus/packages/common/tests/collection-config-store.js, so manual diffing is a bit painful.

File: packages/collection-config-store/tests/test-collection-config-store.js [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+  t.context.version = '6';
+  t.context.collectionConfig = { name: randomString() };
+  t.context.bucket = randomString();
+  // Utility function to return the S3 key of a collection config
+  t.context.collectionConfigKey = (name, version) => {
+    const collectionId = constructCollectionId(name, version);
+    return `${t.context.stackName}/collections/${collectionId}.json`;
+  };
+
+  await createBucket(t.context.bucket);
```
Comment (Jkovarik) [inline packages/collection-config-store/tests/test-collection-config-store.js:30]: Updated  

 await createBucket(t.context.bucket);			      |	  await s3().createBucket({ Bucket: t.context.bucket }).promi

File: packages/collection-config-store/tests/test-collection-config-store.js [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+  const {
+    bucket,
+    collectionConfig,
+    collectionConfigKey,
+    name,
+    version,
+    stackName
+  } = t.context;
+
+  await putJsonS3Object(
```
Comment (Jkovarik) [inline packages/collection-config-store/tests/test-collection-config-store.js:51]: Updated 

```
  await putJsonS3Object(				      |	  await s3().putObject({
    bucket,						      |	    Bucket: bucket,
    collectionConfigKey(name, version),			      |	    Key: collectionConfigKey(name, version),
    collectionConfig					      |	    Body: JSON.stringify(collectionConfig)
  );							      |	  }).promise();
```

File: packages/collection-config-store/tests/test-collection-config-store.js [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
```
Comment (Jkovarik) [inline packages/collection-config-store/tests/test-collection-config-store.js:74]: updated: 

```
  await putJsonS3Object(				      |	  await s3().putObject({
    bucket,						      |	    Bucket: bucket,
    collectionConfigKey(name, version),			      |	    Key: collectionConfigKey(name, version),
    collectionConfig					      |	    Body: JSON.stringify(collectionConfig)
  );							      |	  }).promise();
```

File: packages/collection-config-store/tests/test-collection-config-store.js [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+    collectionConfig,
+    collectionConfigKey,
+    name,
+    version,
+    stackName
+  } = t.context;
+  const collectionConfigStore = new CollectionConfigStore(bucket, stackName);
+  await collectionConfigStore.put(name, version, collectionConfig);
+
+  const storedCollectionConfig = await getJsonS3Object(
```
Comment (Jkovarik) [inline packages/collection-config-store/tests/test-collection-config-store.js:131]: Updated: 

```
  const storedCollectionConfig = await getJsonS3Object(	      |	  const getObjectResponse = await s3().getObject({
    bucket,						      |	    Bucket: bucket,
    collectionConfigKey(name, version)			      |	    Key: collectionConfigKey(name, version)
  );							      |	  }).promise();
							      >
							      >	  const storedCollectionConfig = JSON.parse(getObjectResponse
  t.deepEqual(storedCollectionConfig, collectionConfig);	  t.deepEqual(storedCollectionConfig, collectionConfig);
});								});```

File: packages/collection-config-store/tests/test-collection-config-store.js (hunk 3 above)
Comment (Jkovarik) [reply packages/collection-config-store/tests/test-collection-config-store.js:1]: Updates look good, moved from ../cumulus/packages/common/tests/collection-config-store.js

File: packages/launchpad-auth/tests/test-LaunchpadToken.js [hunk 8]
```diff
@@ -0,0 +1,158 @@
+'use strict';
```
Comment (Jkovarik) [inline packages/launchpad-auth/tests/test-LaunchpadToken.js:1]: moved from common/tests/test-LaunchpadToken.js

Comment (Jkovarik) [inline packages/launchpad-auth/tests/test-LaunchpadToken.js:1]: updated: 

```
'use strict';							'use strict';

const test = require('ava');					const test = require('ava');
							      >	const cryptoRandomString = require('crypto-random-string');
const nock = require('nock');					const nock = require('nock');
const rewire = require('rewire');				const rewire = require('rewire');
const {								const {
							      >	  createBucket,
  deleteS3Object,						  deleteS3Object,
  recursivelyDeleteS3Bucket,					  recursivelyDeleteS3Bucket,
  s3,							      <
  s3PutObject							  s3PutObject
} = require('../aws');					      |	} = require('@cumulus/aws-client/S3');
const { randomString } = require('../test-utils');	      <

const LaunchpadToken = require('../LaunchpadToken');		const LaunchpadToken = require('../LaunchpadToken');
const launchpad = rewire('../launchpad');		      |	const launchpad = rewire('..');
const { getLaunchpadToken, validateLaunchpadToken } = launchp	const { getLaunchpadToken, validateLaunchpadToken } = launchp
const launchpadTokenBucketKey = launchpad.__get__('launchpadT	const launchpadTokenBucketKey = launchpad.__get__('launchpadT

							      >	const randomString = () => cryptoRandomString({ length: 10 })
							      >

```

File: packages/launchpad-auth/tests/test-LaunchpadToken.js [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+  owner_auid: 'testuser',
+  owner_groups: ['cn=cumulus_user_group,ou=Groups,dc=nasa,dc=gov'],
+  session_maxtimeout: 3600,
+  session_starttime: 1564067402,
+  status: 'success'
+};
+
+test.before(async () => {
+  // let's copy the key to s3
+  await createBucket(bucket);
```
Comment (Jkovarik) [inline packages/launchpad-auth/tests/test-LaunchpadToken.js:56]: Updated 

```
  await s3().createBucket({ Bucket: bucket }).promise();      |	  await createBucket(bucket);
```

File: packages/launchpad-auth/tests/test-LaunchpadToken.js (hunk 8 above)
Comment (Jkovarik) [reply packages/launchpad-auth/tests/test-LaunchpadToken.js:1]: updates look good.

File: packages/launchpad-auth/index.js [hunk 10]
```diff
@@ -0,0 +1,148 @@
+'use strict';
```
Comment (Jkovarik) [inline packages/launchpad-auth/index.js:1]: Moved from common/launchpad.js.   Changes look good.   Interesting docstring updates.

Comment (Jkovarik) [reply packages/launchpad-auth/index.js:1]: Full changes:

```
'use strict';									'use strict';

									  >	/**
									  >	 * Utility functions for generating and validating Launchpad tokens
									  >	 *
									  >	 * @module launchpad-auth
									  >	 */
									  >
const path = require('path');							const path = require('path');
const pick = require('lodash/pick');					  |	const pick = require('lodash.pick');
const { getS3Object, s3ObjectExists, s3PutObject } = require('./aws');	  |	const {
									  >	  getS3Object,
									  >	  s3ObjectExists,
									  >	  s3PutObject
									  >	} = require('@cumulus/aws-client/S3');
									  >	const Logger = require('@cumulus/logger');
									  >
const LaunchpadToken = require('./LaunchpadToken');				const LaunchpadToken = require('./LaunchpadToken');
const log = require('./log');						  |
									  >	const log = new Logger({ sender: '@cumulus/launchpad-auth' });

/**										/**
 * get s3 location of the Launchpad token				  |	 * Get S3 location of the Launchpad token
									  >	 *
									  >	 * @returns {Promise.<Object.<string, string>>} - S3 Bucket and Key wh
 *										 *
 * @returns {Promise.<Object.<string, string>>} - s3 Bucket and Key wh	  |	 * @private
 */										 */
function launchpadTokenBucketKey() {						function launchpadTokenBucketKey() {
  if (!(process.env.stackName && process.env.system_bucket)) {			  if (!(process.env.stackName && process.env.system_bucket)) {
    throw new Error('must set environment variables process.env.stackN		    throw new Error('must set environment variables process.env.stackN
  }										  }

  return {									  return {
    Bucket: process.env.system_bucket,						    Bucket: process.env.system_bucket,
    Key: path.join(process.env.stackName, 'launchpad/token.json')		    Key: path.join(process.env.stackName, 'launchpad/token.json')
  };										  };
}										}

/**										/**
 * retrieve Launchpad token from s3					  |	 * Retrieve Launchpad token from S3
 *										 *
 * @returns {Promise.<string>} - the Launchpad token, null if token do		 * @returns {Promise.<string>} - the Launchpad token, null if token do
									  >	 *
									  >	 * @async
									  >	 * @private
 */										 */
async function getValidLaunchpadTokenFromS3() {					async function getValidLaunchpadTokenFromS3() {
  const s3location = launchpadTokenBucketKey();					  const s3location = launchpadTokenBucketKey();
  const keyExists = await s3ObjectExists(s3location);				  const keyExists = await s3ObjectExists(s3location);

  let token = null;								  let token = null;
  if (keyExists) {								  if (keyExists) {
    const s3object = await getS3Object(s3location.Bucket, s3location.K		    const s3object = await getS3Object(s3location.Bucket, s3location.K
    const launchpadToken = JSON.parse(s3object.Body.toString());		    const launchpadToken = JSON.parse(s3object.Body.toString());

    // check if token is still valid						    // check if token is still valid
    if (Date.now() / 1000 < launchpadToken.session_maxtimeout + launch		    if (Date.now() / 1000 < launchpadToken.session_maxtimeout + launch
      token = launchpadToken.sm_token;						      token = launchpadToken.sm_token;
    }										    }
  }										  }

  return token;									  return token;
}										}

/**										/**
 * get Launchpad token							  |	 * Get a Launchpad token
 *										 *
 * @param {Object} params - the configuration parameters for creating 		 * @param {Object} params - the configuration parameters for creating
 * @param {string} params.api - the Launchpad token service api endpoi		 * @param {string} params.api - the Launchpad token service api endpoi
 * @param {string} params.passphrase - the passphrase of the Launchpad		 * @param {string} params.passphrase - the passphrase of the Launchpad
 * @param {string} params.certificate - the name of the Launchpad PKI 		 * @param {string} params.certificate - the name of the Launchpad PKI
 *										 *
 * @returns {Promise.<string>} - the Launchpad token				 * @returns {Promise.<string>} - the Launchpad token
									  >	 *
									  >	 * @async
									  >	 * @alias module:launchpad-auth
 */										 */
async function getLaunchpadToken(params) {					async function getLaunchpadToken(params) {
  let token = await getValidLaunchpadTokenFromS3();				  let token = await getValidLaunchpadTokenFromS3();

  if (!token) {									  if (!token) {
    log.debug('getLaunchpadToken requesting launchpad token');			    log.debug('getLaunchpadToken requesting launchpad token');
    const launchpad = new LaunchpadToken(params);				    const launchpad = new LaunchpadToken(params);
    const tokenObject = await launchpad.requestToken();				    const tokenObject = await launchpad.requestToken();

    // add session_starttime to token object, assume token is generate		    // add session_starttime to token object, assume token is generate
    tokenObject.session_starttime = (Date.now() / 1000 - 60);			    tokenObject.session_starttime = (Date.now() / 1000 - 60);
    const s3location = launchpadTokenBucketKey();				    const s3location = launchpadTokenBucketKey();
    await s3PutObject({								    await s3PutObject({
      Bucket: s3location.Bucket,						      Bucket: s3location.Bucket,
      Key: s3location.Key,							      Key: s3location.Key,
      Body: JSON.stringify(tokenObject)						      Body: JSON.stringify(tokenObject)
    });										    });

    token = tokenObject.sm_token;						    token = tokenObject.sm_token;
  }										  }

  return token;									  return token;
}										}

/**										/**
 * validate Launchpad token						  |	 * Validate a Launchpad token
 *										 *
 * @param {Object} params - the configuration parameters for creating 		 * @param {Object} params - the configuration parameters for creating
 * @param {string} params.api - the Launchpad token service api endpoi		 * @param {string} params.api - the Launchpad token service api endpoi
 * @param {string} params.passphrase - the passphrase of the Launchpad		 * @param {string} params.passphrase - the passphrase of the Launchpad
 * @param {string} params.certificate - the name of the Launchpad PKI 		 * @param {string} params.certificate - the name of the Launchpad PKI
 * @param {string} token - the token to be validated				 * @param {string} token - the token to be validated
 * @param {string} userGroup - the cumulus user group that a valid use		 * @param {string} userGroup - the cumulus user group that a valid use
 *										 *
 * @returns {Promise.<Object>} - the validate result object with		 * @returns {Promise.<Object>} - the validate result object with
 * { status: 'success or failed', message: 'reason for failure',		 * { status: 'success or failed', message: 'reason for failure',
 * session_maxtimeout: number second, session_starttime: number millis		 * session_maxtimeout: number second, session_starttime: number millis
 * owner_auid: string}								 * owner_auid: string}
									  >	 *
									  >	 * @async
									  >	 * @alias module:launchpad-auth
 */										 */
async function validateLaunchpadToken(params, token, userGroup) {		async function validateLaunchpadToken(params, token, userGroup) {
  log.debug('validateLaunchpadToken validating launchpad token');		  log.debug('validateLaunchpadToken validating launchpad token');
  const launchpad = new LaunchpadToken(params);					  const launchpad = new LaunchpadToken(params);
  const response = await launchpad.validateToken(token);			  const response = await launchpad.validateToken(token);
  let result = { status: response.status };					  let result = { status: response.status };

  if (response.status === 'success') {						  if (response.status === 'success') {
    // check if user is in the given group					    // check if user is in the given group
    if (userGroup && userGroup.toUpperCase() !== 'N/A'				    if (userGroup && userGroup.toUpperCase() !== 'N/A'
    && response.owner_groups.filter((group) => group.includes(userGrou		    && response.owner_groups.filter((group) => group.includes(userGrou
      result.status = 'failed';							      result.status = 'failed';
      result.message = 'User not authorized';					      result.message = 'User not authorized';
    }										    }
  } else {									  } else {
    result.message = 'Invalid access token';					    result.message = 'Invalid access token';
  }										  }

  if (result.status === 'success') {						  if (result.status === 'success') {
    const picked = pick(response, ['session_maxtimeout', 'session_star		    const picked = pick(response, ['session_maxtimeout', 'session_star
    result = Object.assign(result, picked);					    result = Object.assign(result, picked);
  }										  }

  return result;								  return result;
}										}

module.exports.getLaunchpadToken = getLaunchpadToken;			  |	module.exports = {
module.exports.validateLaunchpadToken = validateLaunchpadToken;		  |	  getLaunchpadToken,
									  >	  validateLaunchpadToken
									  >	};
```

File: packages/launchpad-auth/index.js [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+ *
+ * @param {Object} params - the configuration parameters for creating LaunchpadToken object
+ * @param {string} params.api - the Launchpad token service api endpoint
+ * @param {string} params.passphrase - the passphrase of the Launchpad PKI certificate
+ * @param {string} params.certificate - the name of the Launchpad PKI pfx certificate
+ *
+ * @returns {Promise.<string>} - the Launchpad token
+ *
+ * @async
+ * @alias module:launchpad-auth
```
Comment (Jkovarik) [inline packages/launchpad-auth/index.js:77]: Hmm, TIL.    Interesting.

File: packages/launchpad-auth/README.md [hunk 12]
```diff
@@ -0,0 +1,19 @@
+# @cumulus/launchpad-auth
+
+Utilities for authentication with Launchpad.
```
Comment (Jkovarik) [inline packages/launchpad-auth/README.md:3]: Nit: 

```suggestion
Utilities for authentication by Cumulus using Launchpad.
```

Or something to specify that this isn't a general-purpose auth lib for Launchpad.    Unless it is.

File: packages/launchpad-auth/LaunchpadToken.js [hunk 13]
```diff
@@ -0,0 +1,155 @@
+'use strict';
```
Comment (Jkovarik) [inline packages/launchpad-auth/LaunchpadToken.js:1]: Moved from packages/common/LaunchpadToken.js

Comment (Jkovarik) [reply packages/launchpad-auth/LaunchpadToken.js:1]: Changes: 

```
'use strict';									'use strict';

const https = require('https');							const https = require('https');
const path = require('path');							const path = require('path');
const { URL } = require('url');							const { URL } = require('url');
const log = require('./log');						  |	const Logger = require('@cumulus/logger');
const { getS3Object, s3ObjectExists } = require('./aws');		  |	const { getS3Object, s3ObjectExists } = require('@cumulus/aws-client/S
									  >
									  >	const log = new Logger({ sender: '@cumulus/launchpad-auth/LaunchpadTok

/**										/**
 * A class for sending requests to Launchpad token service endpoints	  |	 * @class
									  >	 * @classdesc A class for sending requests to Launchpad token service
 *										 *
 * @example									 * @example
 * const LaunchpadToken = require('@cumulus/common/LaunchpadToken');	  |	 * const LaunchpadToken = require('@cumulus/launchpad-auth/LaunchpadTo
 *										 *
 * const LaunchpadToken = new LaunchpadToken({				  |	 * const launchpadToken = new LaunchpadToken({
 *  api: 'launchpad-token-api-endpoint',					 *  api: 'launchpad-token-api-endpoint',
 *  passphrase: 'my-pki-passphrase',						 *  passphrase: 'my-pki-passphrase',
 *  certificate: 'my-pki-certificate.pfx'					 *  certificate: 'my-pki-certificate.pfx'
 * });										 * });
 *										 *
 * @class LaunchpadToken						  |	 * @alias LaunchpadToken
 */										 */
class LaunchpadToken {								class LaunchpadToken {
  /**										  /**
  * @param {Object} params							  * @param {Object} params
  * @param {string} params.api - the Launchpad token service api endpo		  * @param {string} params.api - the Launchpad token service api endpo
  * @param {string} params.passphrase - the passphrase of the Launchpa		  * @param {string} params.passphrase - the passphrase of the Launchpa
  * @param {string} params.certificate - the name of the Launchpad PKI		  * @param {string} params.certificate - the name of the Launchpad PKI
  */										  */
  constructor(params) {								  constructor(params) {
    this.api = params.api;							    this.api = params.api;
    this.passphrase = params.passphrase;					    this.passphrase = params.passphrase;
    this.certificate = params.certificate;					    this.certificate = params.certificate;
  }										  }

  /**										  /**
   * retrieve launchpad credentials					  |	   * Retrieve Launchpad credentials
   *										   *
   * @returns {Promise<Buffer>} - an object with the pfx			   * @returns {Promise<Buffer>} - an object with the pfx
									  >	   * @private
   */										   */
  async _retrieveCertificate() {						  async _retrieveCertificate() {
    if (!(process.env.stackName || process.env.system_bucket)) {		    if (!(process.env.stackName || process.env.system_bucket)) {
      throw Promise.reject(							      throw Promise.reject(
        new Error('must set environment variables process.env.stackNam		        new Error('must set environment variables process.env.stackNam
      );									      );
    }										    }
    const bucket = process.env.system_bucket;					    const bucket = process.env.system_bucket;
    const stackName = process.env.stackName;					    const stackName = process.env.stackName;
    // we are assuming that the specified certificate file is in the S		    // we are assuming that the specified certificate file is in the S

    const cryptKey = `${stackName}/crypto/${this.certificate}`;			    const cryptKey = `${stackName}/crypto/${this.certificate}`;

    const keyExists = await s3ObjectExists(					    const keyExists = await s3ObjectExists(
      { Bucket: bucket, Key: cryptKey }						      { Bucket: bucket, Key: cryptKey }
    );										    );

    if (!keyExists) {								    if (!keyExists) {
      return Promise.reject(new Error(`${this.certificate} does not ex		      return Promise.reject(new Error(`${this.certificate} does not ex
    }										    }

    log.debug(`Reading Key: ${this.certificate} bucket:${bucket},stack		    log.debug(`Reading Key: ${this.certificate} bucket:${bucket},stack
    const pfx = (await getS3Object(bucket, `${stackName}/crypto/${this		    const pfx = (await getS3Object(bucket, `${stackName}/crypto/${this

    return pfx;									    return pfx;
  }										  }

  /**										  /**
   * get token from launchpad						  |	   * Get a token from Launchpad
   *										   *
   * @returns {Promise.<Object>} - the Launchpad gettoken response obj		   * @returns {Promise.<Object>} - the Launchpad gettoken response obj
   */										   */
  async requestToken() {							  async requestToken() {
    log.debug('LaunchpadToken.requestToken');					    log.debug('LaunchpadToken.requestToken');
    const pfx = await this._retrieveCertificate();				    const pfx = await this._retrieveCertificate();
    const launchpadUrl = new URL(this.api);					    const launchpadUrl = new URL(this.api);

    const options = {								    const options = {
      hostname: launchpadUrl.hostname,						      hostname: launchpadUrl.hostname,
      port: launchpadUrl.port || 443,						      port: launchpadUrl.port || 443,
      path: path.join(launchpadUrl.pathname, 'gettoken'),			      path: path.join(launchpadUrl.pathname, 'gettoken'),
      method: 'GET',								      method: 'GET',
      pfx,									      pfx,
      passphrase: this.passphrase						      passphrase: this.passphrase
    };										    };

    const responseBody = await this._submitRequest(options);			    const responseBody = await this._submitRequest(options);
    return JSON.parse(responseBody);						    return JSON.parse(responseBody);
  }										  }

  /**										  /**
   * validate Launchpad token						  |	   * Validate a Launchpad token
   *										   *
   * @param {string} token - the Launchpad token for validation			   * @param {string} token - the Launchpad token for validation
   * @returns {Promise.<Object>} - the Launchpad validate token respon		   * @returns {Promise.<Object>} - the Launchpad validate token respon
   */										   */
  async validateToken(token) {							  async validateToken(token) {
    log.debug('LaunchpadToken.validateToken');					    log.debug('LaunchpadToken.validateToken');
    const pfx = await this._retrieveCertificate();				    const pfx = await this._retrieveCertificate();
    const launchpadUrl = new URL(this.api);					    const launchpadUrl = new URL(this.api);

    const data = JSON.stringify({ token });					    const data = JSON.stringify({ token });
    const options = {								    const options = {
      hostname: launchpadUrl.hostname,						      hostname: launchpadUrl.hostname,
      port: launchpadUrl.port || 443,						      port: launchpadUrl.port || 443,
      path: path.join(launchpadUrl.pathname, 'validate'),			      path: path.join(launchpadUrl.pathname, 'validate'),
      method: 'POST',								      method: 'POST',
      pfx,									      pfx,
      passphrase: this.passphrase,						      passphrase: this.passphrase,
      headers: {								      headers: {
        'Content-Type': 'application/json',					        'Content-Type': 'application/json',
        'Content-Length': data.length						        'Content-Length': data.length
      }										      }
    };										    };

    const responseBody = await this._submitRequest(options, data);		    const responseBody = await this._submitRequest(options, data);
    return JSON.parse(responseBody);						    return JSON.parse(responseBody);
  }										  }

  /**										  /**
   * submit https request						  |	   * Submit HTTPS request
   *										   *
   * @param {Object} options - the Launchpad token for validation		   * @param {Object} options - the Launchpad token for validation
   * @param {string} data - the request body					   * @param {string} data - the request body
   * @returns {Promise.<string>} - the response body				   * @returns {Promise.<string>} - the response body
									  >	   * @private
   */										   */
  _submitRequest(options, data) {						  _submitRequest(options, data) {
    return new Promise((resolve, reject) => {					    return new Promise((resolve, reject) => {
      let responseBody = '';							      let responseBody = '';

      const req = https.request(options, (res) => {				      const req = https.request(options, (res) => {
        if (res.statusCode !== 200) {						        if (res.statusCode !== 200) {
          reject(new Error(`launchpad request failed with statusCode $		          reject(new Error(`launchpad request failed with statusCode $
        }									        }

        res.on('data', (d) => {							        res.on('data', (d) => {
          responseBody += d;							          responseBody += d;
        });									        });

        res.on('end', () => resolve(responseBody));				        res.on('end', () => resolve(responseBody));
      });									      });

      req.on('error', (e) => reject(e));					      req.on('error', (e) => reject(e));

      if (data) req.write(data);						      if (data) req.write(data);
      req.end();								      req.end();
    });										    });
  }										  }
}										}

module.exports = LaunchpadToken;						module.exports = LaunchpadToken;

```

File: packages/collection-config-store/index.js [hunk 14]
```diff
@@ -0,0 +1,134 @@
+'use strict';
```
Comment (Jkovarik) [inline packages/collection-config-store/index.js:1]: Moved from packages/common/collection-config-store.js

Comment (Jkovarik) [reply packages/collection-config-store/index.js:1]: Changes: 

```
'use strict';									'use strict';

const { s3 } = require('./aws');					  <
const { deprecate } = require('./util');				  <
									  <
/**										/**
 * Returns the collectionId used in elasticsearch			  |	 * Utilities for storing and retrieving collection config in S3
 * which is a combination of collection name and version		  <
 *										 *
 * @param {string} name - collection name				  |	 * @module collection-config-store
 * @param {string} version - collection version				  <
 * @returns {string} collectionId					  <
 */										 */
function constructCollectionId(name, version) {				  |
  deprecate('@cumulus/common/collection-config-store.constructCollecti	  |	const {
  return `${name}___${version}`;					  |	  deleteS3Object,
}									  |	  getJsonS3Object,
									  >	  putJsonS3Object
									  >	} = require('@cumulus/aws-client/S3');
									  >	const { constructCollectionId } = require('@cumulus/message/Collection

/**										/**
 * Store and retrieve collection configs in S3				  |	 * @class
									  >	 * @classdesc Store and retrieve collection configs in S3
									  >	 *
									  >	 * @example
									  >	 * const CollectionConfigStore = require('@cumulus/collection-config-s
									  >	 *
									  >	 * const collectionConfigStore = new CollectionConfigStore(
									  >	 *   'system-bucket',
									  >	 *   'stack-name'
									  >	 * );
									  >	 *
									  >	 * @alias module:collection-config-store
 */										 */
class CollectionConfigStore {							class CollectionConfigStore {
  /**										  /**
   * Initialize a CollectionConfigFetcher instance			  |	   * Initialize a CollectionConfigStore instance
   *										   *
   * @param {string} bucket - the bucket where collection configs are 		   * @param {string} bucket - the bucket where collection configs are
   * @param {string} stackName - the Cumulus deployment stack name		   * @param {string} stackName - the Cumulus deployment stack name
   */										   */
  constructor(bucket, stackName) {						  constructor(bucket, stackName) {
    this.bucket = bucket;							    this.bucket = bucket;
    this.stackName = stackName;							    this.stackName = stackName;
    this.cache = {};								    this.cache = {};
  }										  }

  /**										  /**
   * Fetch a collection config from S3 (or cache if available)			   * Fetch a collection config from S3 (or cache if available)
   *										   *
   * @param {string} name - the name of the collection config to fetch		   * @param {string} name - the name of the collection config to fetch
   * @param {string} version - the version of the collection config to		   * @param {string} version - the version of the collection config to
   * @returns {Object} the fetched collection config			  |	   * @returns {Promise<Object>} the fetched collection config
									  >	   *
									  >	   * @async
   */										   */
  async get(name, version) {							  async get(name, version) {
    const collectionId = constructCollectionId(name, version);			    const collectionId = constructCollectionId(name, version);

    // Check to see if the collection config has already been cached		    // Check to see if the collection config has already been cached
    if (!this.cache[collectionId]) {						    if (!this.cache[collectionId]) {
      let response;							  |	      let collectionConfig;
      try {									      try {
        // Attempt to fetch the collection config from S3			        // Attempt to fetch the collection config from S3
        response = await s3().getObject({				  |	        collectionConfig = await getJsonS3Object(this.bucket, this.con
          Bucket: this.bucket,						  <
          Key: this.configKey(collectionId)				  <
        }).promise();							  <
      } catch (err) {								      } catch (err) {
        if (err.code === 'NoSuchKey') {						        if (err.code === 'NoSuchKey') {
          throw new Error(`A collection config for data type "${collec		          throw new Error(`A collection config for data type "${collec
        }									        }

        if (err.code === 'NoSuchBucket') {					        if (err.code === 'NoSuchBucket') {
          throw new Error(`Collection config bucket does not exist: ${		          throw new Error(`Collection config bucket does not exist: ${
        }									        }

        throw err;								        throw err;
      }										      }

      // Store the fetched collection config to the cache			      // Store the fetched collection config to the cache
      this.cache[collectionId] = JSON.parse(response.Body.toString());	  |	      this.cache[collectionId] = collectionConfig;
    }										    }

    return this.cache[collectionId];						    return this.cache[collectionId];
  }										  }

  /**										  /**
   * Store a collection config to S3						   * Store a collection config to S3
   *										   *
   * @param {string} name - the name of the collection config to store		   * @param {string} name - the name of the collection config to store
   * @param {string} version - version of Collection				   * @param {string} version - version of Collection
   * @param {Object} config - the collection config to store			   * @param {Object} config - the collection config to store
   * @returns {Promise<null>} resolves when the collection config has 		   * @returns {Promise<null>} resolves when the collection config has
   *   to S3									   *   to S3
									  >	   *
									  >	   * @async
   */										   */
  async put(name, version, config) {						  async put(name, version, config) {
    const collectionId = constructCollectionId(name, version);			    const collectionId = constructCollectionId(name, version);

    this.cache[collectionId] = config;						    this.cache[collectionId] = config;

    return s3().putObject({						  |	    return putJsonS3Object(
      Bucket: this.bucket,						  |	      this.bucket,
      Key: this.configKey(collectionId),				  |	      this.configKey(collectionId),
      Body: JSON.stringify(config)					  |	      config
    }).promise().then(() => null); // Don't leak implementation detail	  |	    ).then(() => null); // Don't leak implementation details to the ca
  }										  }

  /**										  /**
   * Delete a collection config from S3						   * Delete a collection config from S3
   *										   *
   * @param {string} name - the name of the collection config to delet		   * @param {string} name - the name of the collection config to delet
   * @param {string} version - version of Collection				   * @param {string} version - version of Collection
   * @returns {Promise<null>} resolves when the collection config has 		   * @returns {Promise<null>} resolves when the collection config has
   *   to S3									   *   to S3
									  >	   *
									  >	   * @async
   */										   */
  async delete(name, version) {							  async delete(name, version) {
    const collectionId = constructCollectionId(name, version);			    const collectionId = constructCollectionId(name, version);

    await s3().deleteObject({						  |	    await deleteS3Object(this.bucket, this.configKey(collectionId));
      Bucket: this.bucket,						  <
      Key: this.configKey(collectionId)					  <
    }).promise();							  <

    delete this.cache[collectionId];						    delete this.cache[collectionId];
  }										  }

  /**										  /**
   * Return the S3 key pointing to the collection config			   * Return the S3 key pointing to the collection config
   *										   *
   * @param {string} collectionId - the name and version			   * @param {string} collectionId - the name and version
   * @returns {string} the S3 key where the collection config is locat		   * @returns {string} the S3 key where the collection config is locat
   *										   *
   * @private									   * @private
   */										   */
  configKey(collectionId) {							  configKey(collectionId) {
    return `${this.stackName}/collections/${collectionId}.json`;		    return `${this.stackName}/collections/${collectionId}.json`;
  }										  }
}										}

module.exports.constructCollectionId = constructCollectionId;		  |	module.exports = CollectionConfigStore;
module.exports.CollectionConfigStore = CollectionConfigStore;		  <

```

File: packages/launchpad-auth/index.js (hunk 11 above)
Comment (markdboyd) [reply packages/launchpad-auth/index.js:77]: Yeah, this was necessary for the type of the functions in the generated API docs to be `exported function`, which is what they are

File: CHANGELOG.md (hunk 2 above)
Comment (markdboyd) [reply CHANGELOG.md:12]: Yeah I moved it because it was sitting under the header for release 1.21.0, but this work was merged after that release. The note was in the wrong place because the work was started before the release though

File: CHANGELOG.md (hunk 1 above)
Comment (markdboyd) [reply CHANGELOG.md:29]: Yeah, I had updated these file names previously but forgot to update the CHANGELOG notes
